private/test/src/react/use-check-resource.test.tsx

```diff
@@ -0,0 +1,227 @@
+      principal={{
+        id: "test-user-1",
+        roles: ["role-1", "role-2"],
+      }}
+    >
+      {children}
+    </CerbosProvider>
+  );
+}
+
+function testCerbosHook<TParams>(
+  describeText: string,
+  hook: (params: TParams) => AsyncResult<unknown>,
+  clientFn: Mock,
+  initialParams: TParams,
+  nextParams: TParams,
+): void {
+  describe(describeText, () => {
+    const value = Symbol(describeText);
+
+    resolveAfterWithValue(clientFn, 300, value);
+
+    afterEach(() => {
+      cleanup();
+    });
+
+    beforeEach(() => {
+      vi.clearAllMocks();
+    });
+
+    function render(): ReturnType<
+      typeof renderHook<AsyncResult<unknown>, TParams>
+    > {
+      return renderHook(hook, {
+        wrapper: App,
+        initialProps: initialParams,
+      });
+    }
+
+    it("initially returns loading AsyncState", () => {
+      const { result, unmount } = render();
+
+      expect(clientFn).toHaveBeenLastCalledWith(initialParams, undefined);
+      expect(clientFn).toHaveBeenCalledTimes(1);
+
+      expect(result.current.isLoading).toBe(true);
+      expect(result.current.data).toBeUndefined();
+      expect(result.current.error).toBeUndefined();
+
+      unmount();
+    });
+
+    it("returns data after loading", async () => {
+      const { result } = render();
+
+      await act(async () => await vi.advanceTimersByTimeAsync(300));
+      expect(clientFn).toHaveBeenCalledTimes(1);
+      expect(result.current.isLoading).toBe(false);
+      expect(result.current.data).toBe(value);
+      expect(result.current.error).toBeUndefined();
+    });
+
+    it("avoids unnecessary calls on Cerbos Client when rerendered with the values that match", async () => {
+      const { rerender, result } = render();
+      await act(async () => await vi.advanceTimersByTimeAsync(300));
+
+      rerender(
+        JSON.parse(JSON.stringify(initialParams)) as typeof initialParams,
+      );
+      expect(result.current.isLoading).toBe(false);
+      await act(async () => await vi.advanceTimersByTimeAsync(300));
+      expect(result.current.isLoading).toBe(false);
+      expect(clientFn).toHaveBeenCalledTimes(1);
+    });
+
+    it("calls Cerbos Client with new values when rerendered with different values", async () => {
+      const { rerender, result } = render();
+
+      await act(async () => await vi.advanceTimersByTimeAsync(300));
+
+      rerender(nextParams);
+      expect(clientFn).toHaveBeenLastCalledWith(nextParams, undefined);
+      expect(clientFn).toHaveBeenCalledTimes(2);
+      expect(result.current.isLoading).toBe(true);
+      await act(async () => await vi.advanceTimersByTimeAsync(300));
+      expect(result.current.isLoading).toBe(false);
+      expect(result.current.data).toBe(value);
+      expect(result.current.error).toBeUndefined();
+    });
+
+    it("aborts incomplete Cerbos requests when unmounted", async () => {
+      //TODO
+    });
+  });
+}
+
+async function wait(duration: number): Promise<void> {
```

Comment:
I think you can just `import { setTimeout } from "timers/promises"` (unless that doesn't work with the fake timers?).

Reply:
yep it doesn't work with the fake timers. But I didn't know about this, thanks!